packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1622,6 +1619,49 @@ export class KeyringController extends BaseController<
     return Promise.all(this.#keyrings.map(displayForKeyring));
   }
 
+  /**
+   * Serialize the current array of keyring instances,
+   * including unsupported keyrings by default.
+   *
+   * @param options - Method options.
+   * @param options.includeUnsupported - Whether to include unsupported keyrings.
+   * @returns The serialized keyrings.
+   */
+  async #getSerializedKeyrings(
+    { includeUnsupported }: { includeUnsupported: boolean } = {
+      includeUnsupported: true,
+    },
+  ): Promise<SerializedKeyring[]> {
+    const serializedKeyrings = await Promise.all(
+      this.#keyrings.map(async (keyring) => {
+        const [type, data] = await Promise.all([
+          keyring.type,
+          keyring.serialize(),
+        ]);
+        return { type, data };
+      }),
+    );
+
+    if (includeUnsupported) {
+      serializedKeyrings.push(...this.#unsupportedKeyrings);
+    }
+
+    return serializedKeyrings;
+  }
+
+  /**
+   * Restore a serialized keyrings array.
+   *
+   * @param serializedKeyrings - The serialized keyrings array.
+   */
+  async #restoreSerializedKeyrings(
+    serializedKeyrings: SerializedKeyring[],
+  ): Promise<void> {
+    for (const serializedKeyring of serializedKeyrings) {
+      await this.#restoreKeyring(serializedKeyring);
+    }
```

Comment:
Restoring keyrings without enforcing the order always represented a _risk_, which most likely never turned into an actual bug for this reason:

While _async_ operations might get handed off to the event loop waiting to be resolved, `deserialize` operations on our default set of keyrings are marked as "async", but none of them is a real async operation that can be handed off to the event loop, so it always got executed on the main thread synchronously.

We changed to a `for` loop here to also eliminate the _risk_, in case one of keyrings used deserializes async.

Reply:
Hmm, `init` is also called during keyring restore, but since the HD and Simple keyrings don't have an asynchronous `init` (they don't have an `init` method at all), the above statement should remain valid